cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Object lastNotificationLock;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.lastNotificationLock = new Object();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
+      notificationSubscription = notificationService.subscribe(getFeed(), this, notificationExecutor);
+    }
+
+    @Override
+    public void cancel() {
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(true);
+      }
+      if (notificationSubscription != null) {
+        notificationSubscription.cancel();
+      }
+    }
+
+    /**
+     * Add a new scheduling task based on the data received by the stream referenced by {@code this} object.
+     */
+    public void createScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule, boolean active,
+                                   long baseRunSize, long baseRunTs, boolean persist) {
+      // TODO add a createScheduleTasks, so that if we create multiple schedules for the same stream at the same
+      // time, we don't have to poll the stream many times
+
+      StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+      StreamSizeScheduleTask previous =
+        scheduleTasks.putIfAbsent(getScheduleId(programId, programType, streamSizeSchedule.getName()),
+                                  newTask);
+      if (previous != null) {
+        // We cannot replace an existing schedule - that functionality is not wanted - yet
+        return;
+      }
+
+      // Initialize the schedule task
+      if (baseRunSize == -1 && baseRunTs == -1) {
+        // This is likely to be the first time that we schedule this task - ie it was not in the schedule store
+        // before. Hence we set the base metrics properly
+        long baseTs = System.currentTimeMillis();
+        long baseSize = pollStream();
+        newTask.startSchedule(baseSize, baseTs, active, persist);
+        synchronized (lastNotificationLock) {
+          lastNotification = new StreamSizeNotification(baseTs, baseSize);
+        }
+      } else {
+        newTask.startSchedule(baseRunSize, baseRunTs, active, persist);
```

Comment:
Or can it happen that a task is scheduled, lastBaseSize is != -1, but lastNotification is null? In that case, we would not call received() in lime 337. But the comment there states that we always want to initialize a new task with the current state of the stream size.